kobo/apps/project_ownership/tasks.py

```diff
@@ -258,23 +260,65 @@ def task_restarter():
         minutes=config.PROJECT_OWNERSHIP_RESUME_THRESHOLD
     )
     stuck_threshold = timezone.now() - timedelta(
         minutes=config.PROJECT_OWNERSHIP_STUCK_THRESHOLD
     )
 
-    # Restart tasks that were stopped unintentionally.
-    for transfer_status in (
-        TransferStatus.objects.filter(
-            date_modified__lte=resume_threshold,
-            date_created__gt=stuck_threshold,
-            status=TransferStatusChoices.IN_PROGRESS,
-        )
-        .exclude(status_type=TransferStatusTypeChoices.GLOBAL)
-        .order_by('date_modified')[:settings.MAX_RESTARTED_TASKS]
-    ):
-        async_task.delay(
-            transfer_status.transfer.pk, transfer_status.status_type
-        )
+    tasks_by_transfer = defaultdict(list)
+    full_transfer_pks = set()
+    # Restart tasks that were stopped unintentionally after the transfer started
+    # PENDING tasks were never started so don't consider them "stuck" even if they are
+    # very old
+
+    stopped_in_progress_tasks = Q(
+        date_modified__lte=resume_threshold,
+        date_created__gt=stuck_threshold,
+        status=TransferStatusChoices.IN_PROGRESS,
+    ) & ~Q(status_type=TransferStatusTypeChoices.GLOBAL)
+
+    # Don't start pending attachment tasks independently.
+    # The submission task should take care of it
+    still_pending_tasks = Q(
+        status=TransferStatusChoices.PENDING,
+        date_modified__lte=resume_threshold,
+    ) & ~Q(status_type=TransferStatusTypeChoices.ATTACHMENTS)
+
+    for transfer_status in TransferStatus.objects.filter(
+        stopped_in_progress_tasks | still_pending_tasks
+    ).order_by('date_modified'):
+        transfer_id = transfer_status.transfer_id
+        if transfer_status.status_type == TransferStatusTypeChoices.GLOBAL:
+            full_transfer_pks.add(transfer_id)
+            # remove individual task restarts, if any
+            tasks_by_transfer.pop(transfer_id, None)
+        else:
+            # if we know we'll need to transfer the whole project
+            # don't bother adding individual task restarts
+            if transfer_id in full_transfer_pks:
+                continue
+            tasks_by_transfer[transfer_id].append(transfer_status.status_type)
+
+    full_transfer_pks = list(full_transfer_pks)
+    transfers_to_restart = Transfer.objects.filter(
+        pk__in=full_transfer_pks[: settings.MAX_RESTARTED_TRANSFERS]
+    )
+
+    # transform {'transfer_pk': ['attachments', 'media_files']...}
+    # to [(pk, 'attachments'), (pk, 'media_files')]
+    all_tasks_to_restart = [
+        [(pk, status_type) for status_type in tasks_by_transfer[pk]]
+        for pk in tasks_by_transfer
+    ]
+    all_tasks_to_restart = list(itertools.chain(*all_tasks_to_restart))[
+        : settings.MAX_RESTARTED_TASKS
+    ]
+
+    for transfer_id, status_type in all_tasks_to_restart:
+        async_task.delay(transfer_id, status_type)
+    for transfer in transfers_to_restart:
```

Comment:
What about loading related tables to avoid making more queries to DB and adding more latency. 

```
transfers_to_restart.select_related('invite', 'invite__sender', 'invite__recipient', 'asset')
```

Ideally, I would also defer 'asset.content' if it's not needed. 

```
transfers_to_restart.defer('asset__content').select_related('invite', 'invite__sender', 'invite__recipient', 'asset')
```

Reply:
I'm a little confused. Why would the JOINs necessitate a distinct?